carcara/src/utils.rs

```diff
@@ -166,15 +167,28 @@ impl<K, V> Default for HashMapStack<K, V> {
 
 // TODO: Document this struct
 #[derive(Debug)]
-pub struct Range(Option<usize>, Option<usize>);
+pub struct Range<T = usize>(Option<T>, Option<T>);
 
-impl Range {
-    pub fn contains(&self, n: usize) -> bool {
-        self.0.map_or(true, |bound| n >= bound) && self.1.map_or(true, |bound| n <= bound)
+impl<T: std::cmp::PartialOrd> Range<T> {
+    pub fn contains(&self, n: T) -> bool {
+        self.0.as_ref().map_or(true, |bound| n >= *bound)
+            && self.1.as_ref().map_or(true, |bound| n <= *bound)
     }
 }
 
-impl fmt::Display for Range {
+impl fmt::Display for Range<usize> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Range(Some(a), Some(b)) if a == b => write!(f, "{}", a),
+            Range(Some(a), Some(b)) => write!(f, "between {} and {}", a, b),
+            Range(Some(a), None) => write!(f, "at least {}", a),
+            Range(None, Some(b)) => write!(f, "up to {}", b),
+            Range(None, None) => write!(f, "any number of"),
+        }
+    }
+}
+
+impl fmt::Display for Range<Integer> {
```

Comment:
You can combine these implementations of `fmt::Display` for `Range<usize>` and `Range<Integer>` by using generics:
```suggestion
impl<T: fmt::Display + std::cmp::PartialEq> fmt::Display for Range<T> {
```